File: src/simulations/Payoffs.py
#!/usr/bin/python

import numpy as np

class Payoff:

    # Python constructor
    def __init__(self, obsTime):
        self.obsTime = obsTime

    def at(self, p):
        raise NotImplementedError('Implementation of method at() required.')

    def discountedAt(self, p):
        return self.at(p) / p.numeraire(self.obsTime)
        
    def observationTimes(self):
        return { self.obsTime }  # default implementation

    # simplify payoff scripting

    # normal operators

    def __add__(self, other):
        if isinstance(other,(int,float)):
            other = Fixed(other)
        return Axpy(1.0,self,other)

    def __sub__(self, other):
        if isinstance(other,(int,float)):
            other = Fixed(other)
        return Axpy(-1.0,other,self)
    
    def __mul__(self, other):
        if isinstance(other,(int,float)):
            return Axpy(other,self,None)
        return Mult(self,other)
    
    def __truediv__(self, other):
        if isinstance(other,(int,float)):
            return Axpy(1.0/other,self,None)
        return Div(self,other)

    # reflective operators

    def __radd__(self, other):
        return self.__add__(other)

    def __rsub__(self, other):
        if isinstance(other,(int,float)):
            other = Fixed(other)
        return Axpy(-1.0,self,other)

    def __rmul__(self, other):
        return self.__mul__(other)

    def __rtruediv__(self, other):
        if isinstance(other,(int,float)):
            other = Fixed(other)
        return Div(other,self)

    # logical operators

    def __lt__(self, other):
        if isinstance(other,(int,float)):
            other = Fixed(other)
        return Logical(self,other,'<')

    def __le__(self, other):
        if isinstance(other,(int,float)):
            other = Fixed(other)
        return Logical(self,other,'<=')

    def __eq__(self, other):
        if isinstance(other,(int,float)):
            other = Fixed(other)
        return Logical(self,other,'==')

    def __ne__(self, other):
        if isinstance(other,(int,float)):
            other = Fixed(other)
        return Logical(self,other,'!=')

    def __ge__(self, other):
        if isinstance(other,(int,float)):
            other = Fixed(other)
        return Logical(self,other,'>=')

    def __gt__(self, other):
        if isinstance(other,(int,float)):
            other = Fixed(other)
        return Logical(self,other,'>')

    # set pay time

    def __matmul__(self, other):
        # better check that other is a float
        return Pay(self,other)

    # use payoff as function

    def __call__(self, obsTime):
        raise NotImplementedError('Implementation of call operator () required.')


# basic payoffs

class Fixed(Payoff):
    def __init__(self, x):
        Payoff.__init__(self, 0.0)
        self.x = x
    def at(self, p):
        return self.x
    def __str__(self):
        return '%.4f' % self.x

class Pay(Payoff):
    def __init__(self, x, payTime):
        Payoff.__init__(self, payTime)
        self.x = x
    def at(self, p):
        return self.x.at(p)
    def observationTimes(self):
        return self.x.observationTimes().union({ self.obsTime })
    def __str__(self):
        return '(%s @ %.2f)' % (self.x,self.obsTime)

class Asset(Payoff):
    def __init__(self, t, alias=None):
        Payoff.__init__(self, t)
        self.alias = alias
    def at(self, p):
        return p.asset(self.obsTime,self.alias)
    def __str__(self):
        return '%s(%.2f)' % (self.alias,self.obsTime)

# basic rates payoffs

class ZeroBond(Payoff):
    def __init__(self, t, T, alias=None):
        Payoff.__init__(self, t)
        self.payTime = T
        self.alias = alias
    def at(self, p):
        return p.zeroBond(self.obsTime,self.payTime,self.alias)
    def __str__(self):
        return 'P_%s(%.2f,%.2f)' % (self.alias,self.obsTime,self.payTime)

class LiborRate(Payoff):
    def __init__(self, obsTime, startTime, endTime, yearFraction=None, tenorBasis = 1.0, alias=None):
        Payoff.__init__(self, obsTime)
        self.startTime = startTime
        self.endTime = endTime
        self.yearFraction = yearFraction if yearFraction is not None else (self.endTime - self.startTime)
        self.tenorBasis = tenorBasis
        self.alias = alias
    def at(self, p):
        pStart = p.zeroBond(self.obsTime,self.startTime,self.alias)
        pEnd   = p.zeroBond(self.obsTime,self.endTime,self.alias)
        return (pStart/pEnd*self.tenorBasis - 1.0)/self.yearFraction
    def __str__(self):
        return 'L_%s(%.2f;%.2f,%.2f)' % (self.alias,self.obsTime,self.startTime,self.endTime)

class SwapRate(Payoff):
    def __init__(self, obsTime, floatTimes, floatWeights, fixedTimes, fixedWeights, alias=None):
        Payoff.__init__(self, obsTime)
        self.floatTimes = floatTimes
        self.floatWeights = floatWeights
        self.fixedTimes = fixedTimes
        self.fixedWeights = fixedWeights
        self.alias = alias
    def at(self, p):
        num = sum([ w*p.zeroBond(self.obsTime,T,self.alias) for w,T in zip(self.floatWeights,self.floatTimes) ])
        den = sum([ w*p.zeroBond(self.obsTime,T,self.alias) for w,T in zip(self.fixedWeights,self.fixedTimes) ])
        return num / den
    def __str__(self):
        return 'S_%s(%.2f;%.2f,%.2f)' % (self.alias,self.obsTime,self.floatTimes[0],self.floatTimes[-1])

# arithmetic operations

class Axpy(Payoff):
    def __init__(self, a, x, y=None):
        Payoff.__init__(self, 0.0)
        self.a = a
        self.x = x
        self.y = y
    def at(self, p):
        res = self.a * self.x.at(p)
        if self.y:
            res += self.y.at(p)
        return res
    def observationTimes(self):
        if self.y:
            return self.x.observationTimes().union(self.y.observationTimes())
        return self.x.observationTimes()
    def __str__(self):
        if self.y is None:
            return '%.4f %s' % (self.a,self.x)
        if self.a == 1.0:
            return '(%s + %s)' % (self.x,self.y)
        if self.a == -1.0:
            return '(%s - %s)' % (self.y,self.x)
        return '(%.4f %s + %s)' % (self.a,self.x,self.y)
    
class Mult(Payoff):
    def __init__(self, x, y):
        Payoff.__init__(self, 0.0)
        self.x = x
        self.y = y
    def at(self, p):
        return self.x.at(p) * self.y.at(p)
    def observationTimes(self):
        return self.x.observationTimes().union(self.y.observationTimes())
    def __str__(self):
        return '%s %s' % (self.x,self.y)
    
class Div(Payoff):
    def __init__(self, x, y):
        Payoff.__init__(self, 0.0)
        self.x = x
        self.y = y
    def at(self, p):
        return self.x.at(p) / self.y.at(p)
    def observationTimes(self):
        return self.x.observationTimes().union(self.y.observationTimes())
    def __str__(self):
        return '%s / %s' % (self.x,self.y)

class Max(Payoff):
    def __init__(self, x, y):
        Payoff.__init__(self, 0.0)
        self.x = x
        self.y = y
    def at(self, p):
        return max(self.x.at(p), self.y.at(p))
    def observationTimes(self):
        return self.x.observationTimes().union(self.y.observationTimes())
    def __str__(self):
        return 'Max(%s, %s)' % (self.x,self.y)

class Min(Payoff):
    def __init__(self, x, y):
        Payoff.__init__(self, 0.0)
        self.x = x
        self.y = y
    def at(self, p):
        return min(self.x.at(p), self.y.at(p))
    def observationTimes(self):
        return self.x.observationTimes().union(self.y.observationTimes())
    def __str__(self):
        return 'Min(%s, %s)' % (self.x,self.y)

class Logical(Payoff):
    def __init__(self, x, y, opString):
        Payoff.__init__(self, 0.0)
        self.x = x
        self.y = y
        self.opString = opString
        self.op = None
        if opString == '<':
            self.op = lambda p : float(self.x.at(p)<self.y.at(p))
        if opString == '<=':
            self.op = lambda p : float(self.x.at(p)<=self.y.at(p))
        if opString == '==':
            self.op = lambda p : float(self.x.at(p)==self.y.at(p))
        if opString == '!=':
            self.op = lambda p : float(self.x.at(p)!=self.y.at(p))
        if opString == '>=':
            self.op = lambda p : float(self.x.at(p)>=self.y.at(p))
        if opString == '>':
            self.op = lambda p : float(self.x.at(p)>self.y.at(p))
        if not self.op:
            raise ValueError('Unknown operator: %s.' % opString)
    def at(self, p):
        return self.op(p)
    def observationTimes(self):
        return self.x.observationTimes().union(self.y.observationTimes())
    def __str__(self):
        return '(%s %s %s)' % (self.x,self.opString,self.y)



# Add further payoffs here
#   - exp, log, sqrt
#   - logicals
#   - basket

File: hybmc/simulations/Payoffs.jl

include("../simulations/McSimulation.jl")

abstract type Payoff end
abstract type Leaf <: Payoff end
abstract type BinaryNode <: Payoff end

function at(self::Payoff, p::Path)
    throw(ArgumentError("Implementation of method at() required."))
end

function obsTime(self::Payoff)
    throw(ArgumentError("Implementation of method obsTime() required."))
end

function discountedAt(self::Payoff, p::Path)
    return at(self,p) / numeraire(p,obsTime(self))
end

function discountedAt(self::Payoff, paths::Array{Path,1})
    res = zeros(size(paths,1))
    for j = 1:size(res,1)
        res[j] = at(self,paths[j]) / numeraire(paths[j],obsTime(self))
    end
    return res
end

discountedAt(self::Payoff, sim::McSimulation) = discountedAt(self, paths(sim))

function discountedAt(payoffs::Array, paths::Array{Path,1})
    res = zeros( (size(payoffs,1), size(paths,1)) )
    for j = 1:size(paths,1)
        for i = 1:size(payoffs,1)
            res[i,j] = at(payoffs[i],paths[j]) / numeraire(paths[j],obsTime(payoffs[i]))
        end
    end
    return res
end

discountedAt(payoffs::Array{}, sim::McSimulation) = discountedAt(payoffs, paths(sim))

function observationTimes(self::Payoff)
    throw(ArgumentError("Implementation of method observationTimes() required."))
end

function observationTimes(self::Leaf)
    return Set([ obsTime(self) ])
end

function observationTimes(self::BinaryNode)
    return union(observationTimes(self.x),observationTimes(self.y))
end

function observationTimes(list::Array)
    times = Set{Float64}()
    for item in list
        times = union(times,observationTimes(item))
    end
    return sort([ t for t in times ])
end

# simplify payoff scripting

import Base.+ 
(+)(x::Payoff,y::Payoff) = Axpy(1.0,x,y)
(+)(x::Payoff,y) = Axpy(1.0,x,Fixed(y))
(+)(x,y::Payoff) = Axpy(1.0,Fixed(x),y)
#
import Base.-
(-)(x::Payoff,y::Payoff) = Axpy(-1.0,y,x)
(-)(x::Payoff,y) = Axpy(-1.0,Fixed(y),x)
(-)(x,y::Payoff) = Axpy(-1.0,y,Fixed(x))
#
import Base.*
(*)(x::Payoff,y::Payoff) = Mult(x,y)
(*)(x::Payoff,y) = Mult(x,Fixed(y))
(*)(x,y::Payoff) = Mult(Fixed(x),y)
#
import Base./
(/)(x::Payoff,y::Payoff) = Div(x,y)
(/)(x::Payoff,y) = Div(x,Fixed(y))
(/)(x,y::Payoff) = Div(Fixed(x),y)
#
import Base.%
(%)(x::Payoff,t) = Pay(x,t)
#
import Base.<
(<)(x::Payoff,y::Payoff) = Logical(x,y,(a,b)->Float64(a<b))
(<)(x::Payoff,y) = Logical(x,Fixed(y),(a,b)->Float64(a<b))
(<)(x,y::Payoff) = Logical(Fixed(x),y,(a,b)->Float64(a<b))
import Base.<=
(<=)(x::Payoff,y::Payoff) = Logical(x,y,(a,b)->Float64(a<=b))
(<=)(x::Payoff,y) = Logical(x,Fixed(y),(a,b)->Float64(a<=b))
(<=)(x,y::Payoff) = Logical(Fixed(x),y,(a,b)->Float64(a<=b))
import Base.==
(==)(x::Payoff,y::Payoff) = Logical(x,y,(a,b)->Float64(a==b))
(==)(x::Payoff,y) = Logical(x,Fixed(y),(a,b)->Float64(a==b))
(==)(x,y::Payoff) = Logical(Fixed(x),y,(a,b)->Float64(a==b))
import Base.>=
(>=)(x::Payoff,y::Payoff) = Logical(x,y,(a,b)->Float64(a>=b))
(>=)(x::Payoff,y) = Logical(x,Fixed(y),(a,b)->Float64(a>=b))
(>=)(x,y::Payoff) = Logical(Fixed(x),y,(a,b)->Float64(a>=b))
import Base.>
(>)(x::Payoff,y::Payoff) = Logical(x,y,(a,b)->Float64(a>b))
(>)(x::Payoff,y) = Logical(x,Fixed(y),(a,b)->Float64(a>b))
(>)(x,y::Payoff) = Logical(Fixed(x),y,(a,b)->Float64(a>b))


# basic payoffs

struct Fixed{T<:AbstractFloat} <: Leaf
    x::T
end

at(self::Fixed, p::Path) = self.x
obsTime(self::Fixed) = 0.0

#

struct Pay{T<:AbstractFloat} <: Payoff
    x::Payoff
    payTime::T
end

at(self::Pay, p::Path) = at(self.x, p)
obsTime(self::Pay) = self.payTime
function observationTimes(self::Pay)
    return union(observationTimes(self.x),[self.payTime])
end

#

struct Asset{T<:AbstractFloat} <: Leaf
    obsTime::T
    alias::String
end

Asset(obsTime) = Asset(obsTime,nothing)
at(self::Asset, p::Path) = asset(p,self.obsTime,self.alias)
obsTime(self::Asset) = self.obsTime

#

# basic rates payoffs

struct ZeroBond{T<:AbstractFloat} <: Leaf
    obsTime::T
    payTime::T
    alias::String
end

ZeroBond(obsTime,payTime) = ZeroBond(obsTime,payTime,"")
at(self::ZeroBond, p::Path) = zeroBond(p,self.obsTime,self.payTime,self.alias)
obsTime(self::ZeroBond) = self.obsTime

#

struct LiborRate{T<:AbstractFloat} <: Leaf
    obsTime::T
    startTime::T
    endTime::T
    yearFraction::T
    tenorBasis::T
    alias::String
    _dummy_::Bool  # avoid stack overflow during constructor
end

function LiborRate(obsTime,startTime,endTime;yearFraction=nothing,tenorBasis=1.0,alias=nothing)
    if isnothing(yearFraction)
        yearFraction = endTime - startTime
    end
    if isnothing(alias)
        alias = ""
    end
    return LiborRate(obsTime,startTime,endTime,yearFraction,tenorBasis,alias,true)
end

function at(self::LiborRate, p::Path)
    pStart = zeroBond(p, self.obsTime, self.startTime, self.alias)
    pEnd   = zeroBond(p, self.obsTime, self.endTime, self.alias)
    return (pStart/pEnd*self.tenorBasis - 1.0)/self.yearFraction
end
obsTime(self::LiborRate) = self.obsTime

#

struct SwapRate{T<:AbstractFloat} <: Leaf
    obsTime::T
    floatTimes::T
    floatWeights::T
    fixedTimes::T
    fixedWeights::T
    alias::String
end

SwapRate(obsTime,floatTimes,floatWeights,fixedTimes,fixedWeights) =
    SwapRate(obsTime,floatTimes,floatWeights,fixedTimes,fixedWeights,"")
function at(self::SwapRate, p::Path)
    num = sum([ w*zeroBond(p,self.obsTime,T,self.alias) for (w,T) in zip(self.floatWeights,self.floatTimes) ])
    den = sum([ w*zeroBond(p,self.obsTime,T,self.alias) for (w,T) in zip(self.fixedWeights,self.fixedTimes) ])
    return num / den
end
obsTime(self::SwapRate) = self.obsTime

#

struct FixedLeg{T<:AbstractFloat} <: Leaf
    obsTime::T
    payTimes::T
    payWeights::T
    alias::String
end

FixedLeg(obsTime,payTime,payWeights) = FixedLeg(obsTime,payTime,payWeights,"")
function at(self::FixedLeg, p::Path)
    return sum([ w*zeroBond(p,self.obsTime,T,self.alias) for (w,T) in zip(self.payWeights,self.payTimes) ])
end
obsTime(self::FixedLeg) = self.obsTime

# arithmetic operations

struct Axpy{T<:AbstractFloat} <: BinaryNode
    a::T
    x::Payoff
    y::Payoff
end

at(self::Axpy, p::Path) = self.a * at(self.x,p) + at(self.y,p)
obsTime(self::Axpy) = 0.0

#

struct Mult <: BinaryNode
    x::Payoff
    y::Payoff
end

at(self::Mult, p::Path) = at(self.x,p) * at(self.y,p)
obsTime(self::Mult) = 0.0

#

struct Div <: BinaryNode
    x::Payoff
    y::Payoff
end

at(self::Div, p::Path) = at(self.x,p) / at(self.y,p)
obsTime(self::Div) = 0.0

#

struct Max <: BinaryNode
    x::Payoff
    y::Payoff
end

at(self::Max, p::Path) = max(at(self.x,p),at(self.y,p))
obsTime(self::Div) = 0.0

#

struct Min <: BinaryNode
    x::Payoff
    y::Payoff
end

at(self::Min, p::Path) = min(at(self.x,p),at(self.y,p))
obsTime(self::Div) = 0.0

#

struct Logical <: BinaryNode
    x::Payoff
    y::Payoff
    op
end

at(self::Logical, p::Path) = self.op(at(self.x,p),at(self.y,p))
obsTime(self::Div) = 0.0

